internal/cri/server/podsandbox/sandbox_run_linux.go

```diff
@@ -107,6 +107,11 @@ func (c *Controller) sandboxContainerSpec(id string, config *runtime.PodSandboxC
 		case runtime.NamespaceMode_POD:
 			specOpts = append(specOpts, oci.WithUserNamespace(uids, gids))
 			usernsEnabled = true
+
+			if err := c.pinUserNamespace(id, nsPath); err != nil {
+				return nil, fmt.Errorf("failed to pin user namespace: %w", err)
+			}
+			specOpts = append(specOpts, customopts.WithNamespacePath(runtimespec.UserNamespace, c.getSandboxPinnedUserNamespace(id)))
```

Comment:
I was thinking that I should pin the userns in internal/cri/server package. However, the podsandbox can be remote plugin. If we handle userns in internal/cri/server package, we will have to extend the sandbox interface with new argument `usernsPath`. I'm not sure it's good to add such information in the api, because the sandbox implementation can retrieve user namespace from network namespace. So, I use `ioctl` to fetch the inode and mount it in podsandbox.

https://github.com/containerd/containerd/blob/e8104a48588446fe7e93e2eab5124145c1366cd0/internal/cri/server/sandbox_service.go#L51-L57

https://github.com/containerd/containerd/blob/e8104a48588446fe7e93e2eab5124145c1366cd0/core/sandbox/controller.go#L65-L70

cc @mikebrow @abel-von @mxpv

Reply:
I don't really know what a remote sandbox is, so I'll risk making a question that doesn't make sense. Sorry in advance if that is the case :)

The userns _is_ created when we create the netns now, that is not changing. Why can't we just persist it there too, when we mount the netns? 

I mean, this way doesn't take a param to the userns path either, but it is just being created with the netns. What is the issue to persist it at that point in time too?

Reply:
> The userns is created when we create the netns now, that is not changing. Why can't we just persist it there too, when we mount the netns?

There are two different plugins in containerd.

- `io.containerd.grpc.v1.cri`
- `io.containerd.sandbox.controller.v1`

Currently, the `io.containerd.sandbox.controller.v1` is implementation to setup sandbox environment. The 
`io.containerd.grpc.v1.cri` manages the CNI configuration, which means that it should take responsibility of setting up networking plugin. Based on this, it looks like that `io.containerd.grpc.v1.cri` should take responsibility to pin the user namespace as well.

However, in my opinion, passing the user namespace path through the API is not ideal, similar to the networking namespace. The user namespace path seems redundant when compared to the networking namespace. Therefore, I believe the sandbox implementation should retrieve the user namespace from the networking namespace path.

https://github.com/containerd/containerd/blob/c8b095f3c282914eb6d4657ee0c5c68714380059/api/services/sandbox/v1/sandbox.proto#L101-L105

ping @containerd/committers @containerd/reviewers for more input if it's good to add new param in proto.

Reply:
Makes sense, I'm not familiar (yet) with the new plugins, sorry for the noise :)

The pinning in this function looks odd to me, as this should just create a runtime-spec config. But I'm fine with it, of course, you are the maintainer =)